### ISSUE TITLE
LSF Adapter: Add support all job submission options

### ISSUE THREAD

**Issue by ericfranz** [OSC/ood_core#25]
Most important:

- [x] 1. `Script#native`
- [x] 2. job_environment

Other Script parameters that look like they are supported:

- [x] start_time    `-b [[year:][month:]day:]hour:minute`
- [x] submit_as_hold `-H`
- [x] rerunnable `-r`
- [x] email `-u mail_user` 
    >   -u mail_user
    >               Sends mail to the specified email destination. To
    >               specify a Windows user account, include the domain
    >               name in uppercase letters and use a single
    >               backslash (DOMAIN_NAME\user_name) in
    >               a Windows command line or a double backslash
    >               (DOMAIN_NAME\\user_name) in a UNIX
    >               command line.

- [x] email_on_started `-B`
- [x] email_on_terminated `-N`
- [x] wall_time
    >-W [hour:]minute[/host_name |
    >/host_model]
    >            Sets the runtime limit of the batch job. If a UNIX
    >            job runs longer than the specified run limit, the
    >            job is sent a SIGUSR2 signal, and is killed if it
    >            does not terminate within ten minutes. If a
    >            Windows job runs longer than the specified run
    >            limit, it is killed immediately. (For a detailed
    >            description of how these jobs are killed, see
    >            bkill.)
    >

- [x] error_path `-e`
- [x] output_path `-o`
- [x] input_path `-i input_file` (supports %I and %J in input file name) or `-is`
- [x] priority `-sp priority` (integer 1 - MAX_USER_PRIORITY)
- [x] queue_name
- [x] reservation_id: `-U reservation_ID`

These don't look like they are supported

- args
- join_files
- min_phys_memory `-M [MB]` claims rosetta; bjobs man page for 8.3 says:
    >-M mem_limit
    > Sets a per-process (soft) memory limit for all the
    > processes that belong to this batch job (see
    > getrlimit(2)).
    > 
    > By default, the limit is specified in KB. Use
    > LSF_UNIT_FOR_LIMITS in lsf.conf to specify a
    > larger unit for the limit (MB, GB, TB, PB, or EB).

**Comment by ericfranz** [comment]
https://slurm.schedmd.com/rosetta.pdf is useful

**Comment by ericfranz** [comment]
The `Script#min_phys_memory` documentation states: 

>The minimum amount of physical memory in kilobyte

Unfortunately, I don't think the -M option allows us to specify the unit. Rather it is specified in LSF_UNIT_FOR_LIMITS in lsf.conf, which can vary from site to site. So for the LSF adapter `Script#min_phys_memory` may be in a unit other than kilobyte.

Also, FWIW:

>-M mem_limit
>Sets a per-process (soft) memory limit for all the
>processes that belong to this batch job (see
>getrlimit(2)).

This actually seems to be the opposite of what `Script#min_phys_memory` is supposed to do.

So, not supporting `Script#min_phys_memory` at this time.

**Comment by nickjer** [comment]
If we feel this cannot be generalized across all resource managers, then this should be a candidate to be removed.

For example in PBS/Torque this is the total memory across all nodes, for Slurm this is the memory per node, and it seems for LSF this may be the memory per process.

**Comment by ericfranz** [comment]
Another thing, in LSF it is memory limit (max mem) not min memory required (min mem).

**Comment by nickjer** [comment]
For job environment is this still relevant:

https://www.ibm.com/support/knowledgecenter/SSETD4_9.1.3/lsf_command_ref/bsub.env.1.html

**Comment by ericfranz** [comment]
That option is not available in 8.3.

FWIW 8.3 manpages:

> bsub does not start a login shell but runs the job file under the execution environment from which the job was submitted

**Comment by ericfranz** [comment]
Setting the env when forking seems to work.

```
hello = <<-EOF
!/bin/bash
#
#BSUB -q short # queue
#BSUB -e myjob.%J.err
#BSUB -o myjob.%J.out
#BSUB -J helloworld

echo "Hello world, I am in $PWD"
echo "This is the value of MY_CUSTOM_VAR: $MY_CUSTOM_VAR"
sleep 30
echo "Goodbye, world!"
EOF

puts adapter.submit(OodCore::Job::Script.new(content: hello, job_environment: { "MY_CUSTOM_VAR" => "123456" }))
```

and the output included:

```
The output (if any) follows:

Hello world, I am in /home/efranz/ood_core
This is the value of MY_CUSTOM_VAR: 123456
Goodbye, world!
```
